Corrupted FS when writing to SD Card

executing this code has repeatedly resulted in a corrupted FS with teh current master CP 3.0
It often fails to allocate enough memory, but sometime runs. If it does all seem ok for awhile -then the FS seem to just disappear.... 
I have reproduced this on a feather_m0_express and on a feather_m0_supersized.
Both with a featherwing Adalogger attached.

```
# Initialize and mount SD card of Featherboard adalogger
import board
import busio
import digitalio
import adafruit_pcf8523
import adafruit_sdcard
import storage
import adafruit_dht

# create the SPI bus and a digital output for the microSD card's c\s line
spi = busio.SPI(board.SCK, MOSI=board.MOSI, MISO=board.MISO)
cs = digitalio.DigitalInOut(board.D5)


# create the microSD card object and the filesystem object
sdcard = adafruit_sdcard.SDCard(spi, cs)
vfs = storage.VfsFat(sdcard)

# mount the microSD card's filesystem into the CircuitPython filesystem
storage.mount(vfs, "/sd")

# Now ready to read and write data from the sd card
with open("/sd/test.txt", "w") as f:
    f.write("Hello world!\r\n")

# open a file and read a line from it with similar code
with open("/sd/test.txt", "r") as f:
    print("Read line from file:")
    print(f.readline())

```

This is the strangest thing.

Looks like I've found a reproduceable way of flipping back and forth between working and not working.
I have a build script that creates a git tag when checking out the code so I can easily dump out the commits I've done using ```git format-patch patchbase..HEAD```.


So with that tag I get a firmware that can handle power cycling:

git tagging:
```
pi@agl:~/circuitpython/workdirs/test/circuitpython$ git tag patchbase
pi@agl:~/circuitpython/workdirs/test/circuitpython$ git log --oneline -1
1eb412b44 (HEAD -> test, tag: patchbase, origin/master, origin/HEAD, master) Merge pull request #845 from matt-land/feature-default-serial
```

(re)Build:
```
pi@agl:~/circuitpython$ rake compile
cd /home/pi/circuitpython/workdirs/test/circuitpython/ports/atmel-samd
make BOARD=feather_m0_express CROSS_COMPILE=/home/pi/opt/gcc-arm-none-eabi-7-2017-q4-major/bin/arm-none-eabi-
Use make V=1, make V=2 or set BUILD_VERBOSE similarly in your environment to increase build verbosity.
Generating build-feather_m0_express/genhdr/mpversion.h
QSTR not updated
../../shared-bindings/usb_hid/__init__.h:33:23: warning: size of 'common_hal_usb_hid_devices' differ from the size of original declaration [-Wlto-type-mismatch]
 extern mp_obj_tuple_t common_hal_usb_hid_devices;
                       ^
common-hal/usb_hid/__init__.c:127:16: note: 'common_hal_usb_hid_devices' was previously declared here
 mp_obj_tuple_t common_hal_usb_hid_devices = {
                ^

33932 bytes free in flash out of 253696 bytes ( 247.75 kb ).
3936 bytes free in ram for stack out of 32768 bytes ( 32.0 kb ).

Create build-feather_m0_express/firmware.bin
Create build-feather_m0_express/firmware.uf2
../../tools/uf2/utils/uf2conv.py -b 0x2000 -c -o build-feather_m0_express/firmware.uf2 build-feather_m0_express/firmware.bin
Converting to uf2, output size: 439808, start address: 0x2000
Wrote 439808 bytes to build-feather_m0_express/firmware.uf2.
cd -
```

Board REPL, the tag show up:
```
Adafruit CircuitPython patchbase-dirty on 2018-05-18; Adafruit Feather M0 Express with samd21g18
>>>
```


Next I remove the tag and rebuild, and it can't handle power cycling:
```
pi@agl:~/circuitpython/workdirs/test/circuitpython$ git tag -d patchbase
Deleted tag 'patchbase' (was 1eb412b44)
pi@agl:~/circuitpython/workdirs/test/circuitpython$ git log --oneline -1
1eb412b44 (HEAD -> test, origin/master, origin/HEAD, master) Merge pull request #845 from matt-land/feature-default-serial
```

Build:
```
pi@agl:~/circuitpython$ rake compile
cd /home/pi/circuitpython/workdirs/test/circuitpython/ports/atmel-samd
make BOARD=feather_m0_express CROSS_COMPILE=/home/pi/opt/gcc-arm-none-eabi-7-2017-q4-major/bin/arm-none-eabi-
Use make V=1, make V=2 or set BUILD_VERBOSE similarly in your environment to increase build verbosity.
Generating build-feather_m0_express/genhdr/mpversion.h
QSTR not updated
../../shared-bindings/usb_hid/__init__.h:33:23: warning: size of 'common_hal_usb_hid_devices' differ from the size of original declaration [-Wlto-type-mismatch]
 extern mp_obj_tuple_t common_hal_usb_hid_devices;
                       ^
common-hal/usb_hid/__init__.c:127:16: note: 'common_hal_usb_hid_devices' was previously declared here
 mp_obj_tuple_t common_hal_usb_hid_devices = {
                ^

33884 bytes free in flash out of 253696 bytes ( 247.75 kb ).
3936 bytes free in ram for stack out of 32768 bytes ( 32.0 kb ).

Create build-feather_m0_express/firmware.bin
Create build-feather_m0_express/firmware.uf2
../../tools/uf2/utils/uf2conv.py -b 0x2000 -c -o build-feather_m0_express/firmware.uf2 build-feather_m0_express/firmware.bin
Converting to uf2, output size: 439808, start address: 0x2000
Wrote 439808 bytes to build-feather_m0_express/firmware.uf2.
cd -
```

REPL, no tag instead commit ref:
```
Adafruit CircuitPython 3.0.0-alpha.6-142-g1eb412b44-dirty on 2018-05-18; Adafruit Feather M0 Express with samd21g18
>>>
```

I have tried writing the failing firmware 3 times without any changes. I've checked that ```boot_out.txt``` gets updated.

Can it be some kind of offset problem? The flash size changes for me between working and not working.

What about this warning?
```
../../shared-bindings/usb_hid/__init__.h:33:23: warning: size of 'common_hal_usb_hid_devices' differ from the size of original declaration [-Wlto-type-mismatch]
 extern mp_obj_tuple_t common_hal_usb_hid_devices;
                       ^
common-hal/usb_hid/__init__.c:127:16: note: 'common_hal_usb_hid_devices' was previously declared here
 mp_obj_tuple_t common_hal_usb_hid_devices = {
```

--- a/main.c
+++ b/main.c
@@ -329,8 +329,8 @@ int __attribute__((used)) main(void) {
         if (!skip_boot_output) {
             f_close(boot_output_file);
             filesystem_flush();
-            boot_output_file = NULL;
         }
+        boot_output_file = NULL;
         #endif
 
         // Reset to remove any state that boot.py setup. It should only be used to

